// react-sliding-container/lib/helpers/circular.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNext = function (list, position) {
    if (list.length > 2 && position === list.length - 1) {
        return list[0];
    }
    else if (list.length > 2) {
        return list[position + 1];
    }
    else if (list.length === 2 && position === 0) {
        return list[position + 1];
    }
    return null;
};
exports.getNextIndex = function (list, position) {
    if (list.length > 2 && position === list.length - 1) {
        return 0;
    }
    else if (list.length > 2) {
        return position + 1;
    }
    else if (list.length === 2 && position === 0) {
        return position + 1;
    }
    return position;
};
exports.getPrevious = function (list, position) {
    if (list.length > 2 && position === 0) {
        return list[list.length - 1];
    }
    else if (list.length > 2) {
        return list[position - 1];
    }
    else if (list.length === 2 && position === list.length - 1) {
        return list[position - 1];
    }
    return null;
};
exports.getPreviousIndex = function (list, position) {
    if (list.length > 2 && position === 0) {
        return list.length - 1;
    }
    else if (list.length > 2) {
        return position - 1;
    }
    else if (list.length === 2 && position === list.length - 1) {
        return position - 1;
    }
    return position;
};
exports.circular = function (list, position) {
    if (position === void 0) { position = 0; }
    var previous = exports.getPrevious(list, position);
    var current = list[position];
    var next = exports.getNext(list, position);
    var elements = {
        calculatedChildren: [
            previous,
            current,
            next
        ],
        current: current,
        length: list.length,
        next: next,
        nextIndex: exports.getNextIndex(list, position),
        position: position,
        previous: previous,
        previousIndex: exports.getPreviousIndex(list, position)
    };
    return elements;
};
